plugin/src/HeadlessAPI.lua

```diff
@@ -0,0 +1,452 @@
+		for i = 1, args.n do
+			local v = args[i]
+			local t = type(v)
+			if t == "string" then
+				args[i] = string.format("%q", v)
+			else
+				args[i] = tostring(v)
+			end
+		end
+
+		print(
+			string.format(
+				"Rojo:Test(%s) called from '%s' (%s)",
+				table.concat(args, ", "),
+				Rojo:_getCallerName(),
+				Rojo:_getCallerSource()
+			)
+		)
+	end
+
+	Rojo._apiDescriptions.ConnectAsync = "Connects to a Rojo server"
+	function Rojo:ConnectAsync(host: string?, port: string?)
+		assert(type(host) == "string" or host == nil, "Host must be type `string?`")
+		assert(type(port) == "string" or port == nil, "Port must be type `string?`")
+
+		if Rojo:_checkRateLimit("ConnectAsync") then
+			return
+		end
+
+		app:startSession(host, port)
+	end
+
+	Rojo._apiDescriptions.DisconnectAsync = "Disconnects from the Rojo server"
+	function Rojo:DisconnectAsync()
+		if Rojo:_checkRateLimit("DisconnectAsync") then
+			return
+		end
+
+		app:endSession()
+	end
+
+	Rojo._apiDescriptions.GetSetting = "Gets a Rojo setting"
+	function Rojo:GetSetting(setting: string): any
+		assert(type(setting) == "string", "Setting must be type `string`")
+
+		return Settings:get(setting)
+	end
+
+	Rojo._apiDescriptions.SetSetting = "Sets a Rojo setting"
+	function Rojo:SetSetting(setting: string, value: any)
+		assert(type(setting) == "string", "Setting must be type `string`")
+
+		if Rojo:_checkRateLimit("SetSetting") then
+			return
+		end
+
+		return Settings:set(setting, value)
+	end
+
+	Rojo._apiDescriptions.Notify = "Shows a notification in the Rojo UI"
+	function Rojo:Notify(
+		msg: string,
+		timeout: number?,
+		actions: { [string]: { text: string, style: string, layoutOrder: number, onClick: (any) -> () } }?
+	): () -> ()
+		assert(type(msg) == "string", "Message must be type `string`")
+		assert(type(timeout) == "number" or timeout == nil, "Timeout must be type `number?`")
+		assert((actions == nil) or (type(actions) == "table"), "Actions must be table or nil")
+
+		if Rojo:_checkRateLimit("Notify") then
+			return function() end
+		end
+
+		local sanitizedActions = nil
+		if actions then
+			sanitizedActions = {}
+			for id, action in actions do
+				assert(type(id) == "string", "Actions key must be string")
+				local actionId = "Actions." .. id
+				assert(type(action) == "table", actionId .. " must be table")
+				assert(type(action.text) == "string", actionId .. ".text must be string")
+				assert(type(action.style) == "string", actionId .. ".style must be string")
+				assert(
+					action.style == "Solid" or action.style == "Bordered",
+					actionId .. ".style must be 'Solid' or 'Bordered'"
+				)
+				assert(type(action.layoutOrder) == "number", actionId .. ".layoutOrder must be number")
+				assert(type(action.onClick) == "function", actionId .. ".onClick must be function")
+
+				sanitizedActions[id] = {
+					text = action.text,
+					style = action.style,
+					layoutOrder = action.layoutOrder,
+					onClick = action.onClick,
```

Comment:
It feels like a potential security risk to directly pass the function like this, since it means a malicious actor *could* get our internal environment.

I don't know if this is something worth worrying about due to the ModuleScript design this uses (you can use `getfenv(Rojo.API.Notify)` to a similar effect) but it's worth thinking about.

Reply:
I can wrap it in an anon function that calls the third party's `onClick` through a `task.spawn` to orphan it a little more.

`onClick = function() print(debug.traceback()) end`
As is:
> user_TestingRojoApi.rbxmx.TestingRojoApi.Script:11 function onClick
> user_Rojo.rbxm.Rojo.Plugin.App.Notifications:100 function onClick

Wrapped with spawn:
> user_TestingRojoApi.rbxmx.TestingRojoApi.Script:11 function onClick